Add a link completer for inline links to posts

draganescu opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
In the block editor we currently have the `/` inserter which is an autocompleter for blocks. This PR adds a new autcompleter triggered by `[[` with links to existing posts and pages.

## How has this been tested?
For now tested manually.

## Screenshots <!-- if applicable -->
https://user-images.githubusercontent.com/107534/154686186-ad46a3f5-f94d-4c7b-ac88-39f9b2a2cf8c.mp4


## Types of changes
Non breaking change. Adds a new autocompleter to the `block-editor` package.


## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

Commit:
adds a link completer for inline link to post creation

jasmussen commented:
This is what I see:

![autocomplete](https://user-images.githubusercontent.com/1204802/108684671-7a243400-74f3-11eb-9ba4-c7bca7ae69f1.gif)

_Note that the positioning of the dropdown appears unrelated to this PR (see https://github.com/WordPress/gutenberg/pull/28917#issuecomment-781996219)._

My instinct here: this is really nice! It doesn't look like a lot of code complexity, but it's a really nice little feature for linking to existing stuff. It's so rare that I'd ever want to write `[[` (which is why it's a standard, I suppose), and if I do want it, I can just use the arrow keys. It takes a ton of vinegar out of linking to pages. 

Moreover, it doesn't weigh down the user interface, because it's just a "commandline tool", and it shares such DNA with the slash command already. I personally think there's a place for this. 

If others are fans, I'd tweak a few small things:

- We should add post or page icons depending on the type. This would beneefit if we in the future wanted to also allow linking to tags and categories.
- If a post or page title is empty (as is shown in the GIF above), can we just omit it from this command? You wouldn't be able to search for it anyway.

Nice work!

draganescu commented:
Rebased to fix the wrong placement for the inserter.

Commit:
adds icons and removes blank title items

draganescu commented:
@jasmussen finally I found some time to come back to this. In this state:

- I added icons, but I could not find a "post" icon, while we have a "page" icon, so I used 'customPostType' ...
- I removed blank hits

ironprogrammer commented:
Works as advertised! @draganescu, this is a great use of `[[` to simplify internal linking ([an important part of SEO](https://moz.com/learn/seo/internal-link)).

### Testing Notes
- Safari 15.3 (dark mode by Noir extension)
- WordPress 6.0-alpha-52448-src
- Gutenberg 12.6.0

In the screencast below, I covered the following scenarios:
- Add link to page, use arrows to find.
- Add link to post, use text search to find.
- Use Escape key to dismiss popup.
- Use arrows to dismiss popup.
- Try a shortcode to ensure there is no interference.

https://user-images.githubusercontent.com/824344/154761277-8816190f-76f9-41b3-85f0-8fecfef3158a.mp4

jasmussen commented:
Took it for a spin:
![linking](https://user-images.githubusercontent.com/1204802/154928421-f79ad11b-c37c-4ce8-b4c7-cec00bb52a5d.gif)

Seems cool! It's a pretty power-user specific feature, but since it doesn't weigh the UI at all, it's a nice balance. If it doesn't affect performance negatively, it seems fine for this one to get a code review and land as is! 

We could do a few improvements in the nitpick category, though. It'd be nice to add this new keyboard shortcut to the shortcuts sheet, this one:
<img width="754" alt="shortcuts" src="https://user-images.githubusercontent.com/1204802/154928802-164b5c81-c1d6-4400-9204-c65350336b42.png">

It also seems like we should have an icon for posts. I don't really love the "Pin" for posts, even if it has historically been represented that way. Nevertheless we do have a pin icon in the library:
<img width="375" alt="Screenshot 2022-02-21 at 10 41 30" src="https://user-images.githubusercontent.com/1204802/154929929-5e77d1d5-f688-4a83-b17d-d71d4c310027.png">

We could also, in the name of post types, use the same general pattern, and do this:
<img width="390" alt="Screenshot 2022-02-21 at 10 47 27" src="https://user-images.githubusercontent.com/1204802/154929975-e0cb9976-6901-433e-8d6a-28e4e2669db0.png">

None of those icons are published yet, but we probably want them. Let me know if you want the vectors to add to the library.

jasmussen commented:
I wish I'd get notified for thumbs up 😅 — sorry for the delay. Here are new icons: https://github.com/WordPress/gutenberg/pull/39139

Commit:
updated icon for post and added shortcut to modal

draganescu commented:
Added the new icon and the shortcut in the help modal. Something I tried and failed is to timeout the sending of requests but for some reason using `debounce` breaks the suggestion list (it works but slower but it is not optimized for this kind of return). I suggest we merge like this and update the autocomplete thing to Suport denouncing (not sending one request per keystroke).

jasmussen reviewed (approved):
Giving tentative approval to this one because:

* It's a fun power user feature that is likely to make a few people very happy
* It does not add any UI weight virtualy at all since it's likely to be rare people intentionally type `[[` for typographic reasons

It also works well:

![autocomplete](https://user-images.githubusercontent.com/1204802/156318700-dac77f8d-22f8-458e-9fad-083e74149177.gif)

One small thing, the keyboard shortcut help looks good (see the bottom):
<img width="576" alt="Screenshot 2022-03-02 at 08 57 11" src="https://user-images.githubusercontent.com/1204802/156319130-d4bbfcb5-b83c-401d-b066-f08d2957d1b6.png">

But shouldn't it be under "Block shortcuts"? Perhaps right under the slash command since its behavior is somewhat similar? I know the command isn't block level so it's a stretch. But it works _in a block_? Alternately we could rename the "Text formatting" to be "Text formatting & shortcuts"?

It'd also be nice to get another code sanity check, but otherwise: nice one!

draganescu commented:
@jasmussen thanks for the 👍🏻 The shortcut is very related to the link formatters so I'd leave it under "Text formatting", but move it up next to the other link format shortcuts.

Commit:
update snapshot and place the shortcut near other link shortcuts

nickchomey commented:
I just want to say that I **deeply** appreciate this feature, which I just discovered with Gutenberg 12.8. I had been futzing with various plugins that did a poor job of this sort of thing. I think you'll be very surprised at how many non-"power users" will love this as well! Thanks again!

Edit: One point of feedback - it appears to work for CPTs, but not for bbPress Forums, Discussions or Replies, despite the fact that they are also CPTs. Is that something that could be investigated/added? I could see it being very useful. Perhaps a setting somewhere to allow us to select which CPTs can be autolinked?

westonruter commented:
Just a thought: why not using `@` instead of `[[`? This would align with how other platforms provide means of tagging other entities, not just links. Google Workspace for example now uses `@` to bring up a menu to [insert smart chips](https://support.google.com/docs/answer/10710316?hl=en). Of course GitHub, Slack, etc use `@` for mentioning specific users. Seems `@` would be more discoverable than `[[`. Maybe it could be used a general inline token inserter menu (e.g. images, current date) rather than just for link insertion.

Mamaduka commented:
@westonruter the editor also uses `@` for user autocompletion.

nickchomey commented:
@westonruter The `[[ ]]` format surely comes from wikipedia. There are also various notetaking/knowledge management tools that are gaining popularity, like Obsidian, Roam and Athens, that use `[[ ]]`. 

I'm ultimately indifferent to what is used, so long as it is functional. My concern would be confusion since, as @Mamaduka says, @ is already used for user autocompletion. 

`@` for users, `/` for blocks and `[[ ]]` for links seems appropriate

johnfleming commented:
Test cases all seem to work. Thumbs up to adding the ability to link to categories and tags
